### PR TITLE
feat: exporting cache classes

### PR DESCRIPTION
**Pull request by erickzanardo**
# Description

Flame has a couple of useful classes that can be used to cache stuff in memory, but they were internal and not exported.

I plan on using these classes to help with perfomance on the flame svg package, so this PR simply move those classes to a more generic folder and export them and fix the places that was importing those classes.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the
relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have read the [Contributor Guide] and followed the process outlined for submitting PRs.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples`.

## Breaking Change

Does your PR require Flame users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change. (Indicate it in the [Conventional Commit] prefix with a `!`,
  e.g. `feat!:`, `fix!:`).
- [x] No, this is *not* a breaking change.

## Related Issues

*Replace this paragraph with a list of issues related to this PR from the [issue database].
Indicate which of these issues are resolved or fixed by this PR, i.e. Fixes #xxxx*

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md
[Conventional Commit]: https://conventionalcommits.org

### CODE REVIEW

**Review by st-pasha — COMMENTED** [review]
1. Should the AssetsCache class and Images also be moved into the `cache` folder?

2. Looks like there is a bug in MemoryCache: when `setValue` is called but the key is already there, then the value won't be set.

File: packages/flame/lib/src/cache/cache.dart [hunk 1]
```diff
@@ -0,0 +1,2 @@
+export 'memory_cache.dart';
+export 'value_cache.dart';
```
Comment (spydon) [inline packages/flame/lib/src/cache/cache.dart:2]: Is this file needed? Can't they be exported directly in the other file?

**Review by spydon — APPROVED** [review]

**Review by luanpotter — APPROVED** [review]
I'm always weary of hiding things from users, you never know what can be useful!

File: packages/flame/lib/src/cache/cache.dart (hunk 1 above)
Comment (erickzanardo) [reply packages/flame/lib/src/cache/cache.dart:2]: They could, but I think it is good to have this barrel file, so we can use it internally.

I don't mind removing it if you feel too strong about it though.

Comment (spydon) [reply packages/flame/lib/src/cache/cache.dart:2]: I don't know how it usually is done, but can't we just use the other exported file internally?

**Comment by erickzanardo** [comment]
> 1. Should the AssetsCache class and Images also be moved into the `cache` folder?
> 2. Looks like there is a bug in MemoryCache: when `setValue` is called but the key is already there, then the value won't be set.

1. I don't think so, they are more related to assets than cache
2. I think we did that intentionally, can't remember why now, if anyone feel this needs to be changed it could be addressed in a different PR

File: packages/flame/lib/src/cache/cache.dart (hunk 1 above)
Comment (erickzanardo) [reply packages/flame/lib/src/cache/cache.dart:2]: Yeah, but having a barrel we can import everything related to cache on a single import, otherwise we will need to individually import each file

Comment (spydon) [reply packages/flame/lib/src/cache/cache.dart:2]: If the exports moved to `packages/flame/lib/cache.dart` and this file removed, wouldn't it have the same effect?

Comment (spydon) [reply packages/flame/lib/src/cache/cache.dart:2]: Or shouldn't the externally exported files be used internally? There is something I'm missing I think. :sweat_smile:

Comment (erickzanardo) [reply packages/flame/lib/src/cache/cache.dart:2]: > If the exports moved to `packages/flame/lib/cache.dart` and this file removed, wouldn't it have the same effect?

Yes! And I would honestly prefer that. But we have been doing relatively imports internationally, so that is why I created this internal barrel

**Comment by st-pasha** [comment]
>> Should the AssetsCache class and Images also be moved into the cache folder?
>
> I don't think so, they are more related to assets than cache

Still, isn't it a **cache** of assets, or a **cache** of images?

I personally find directory name `assets` to be somewhat confusing, because in all other cases that directory would contain the assets of the project, i.e. images, sprites, animations, etc. But in our case the `assets` directory contains code, so it's confusing.
